bottlecap/src/metrics/enhanced/lambda.rs

```diff
@@ -429,6 +429,115 @@ impl Lambda {
         });
     }
 
+    pub fn generate_fd_enhanced_metrics(
+        fd_max: f64,
+        fd_use: f64,
+        aggr: &mut std::sync::MutexGuard<Aggregator>,
+    ) {
+        let metric = Metric::new(
+            constants::FD_MAX_METRIC.into(),
+            MetricValue::distribution(fd_max),
+            None,
+        );
+        if let Err(e) = aggr.insert(metric) {
+            error!("Failed to insert fd_max metric: {}", e);
+        }
+
+        // Check if fd_use value is valid before inserting metric
+        if fd_use > 0.0 {
+            let metric = Metric::new(
+                constants::FD_USE_METRIC.into(),
+                MetricValue::distribution(fd_use),
+                None,
+            );
+            if let Err(e) = aggr.insert(metric) {
+                error!("Failed to insert fd_use metric: {}", e);
+            }
+        }
+    }
+
+    pub fn generate_threads_enhanced_metrics(
+        threads_max: f64,
+        threads_use: f64,
+        aggr: &mut std::sync::MutexGuard<Aggregator>,
+    ) {
+        let metric = Metric::new(
+            constants::THREADS_MAX_METRIC.into(),
+            MetricValue::distribution(threads_max),
+            None,
+        );
+        if let Err(e) = aggr.insert(metric) {
+            error!("Failed to insert threads_max metric: {}", e);
+        }
+
+        // Check if threads_use value is valid before inserting metric
+        if threads_use > 0.0 {
+            let metric = Metric::new(
+                constants::THREADS_USE_METRIC.into(),
+                MetricValue::distribution(threads_use),
+                None,
+            );
+            if let Err(e) = aggr.insert(metric) {
+                error!("Failed to insert threads_use metric: {}", e);
+            }
+        }
+    }
+
+    pub fn set_process_enhanced_metrics(&self, mut send_metrics: Receiver<()>) {
+        if !self.config.enhanced_metrics {
+            return;
+        }
+
+        let aggr = Arc::clone(&self.aggregator);
+
+        tokio::spawn(async move {
+            // get list of all process ids
+            let pids = proc::get_pid_list();
+
+            // Set fd_max and initial value for fd_use to -1
+            let fd_max = proc::get_fd_max_data(&pids);
+            let mut fd_use = -1_f64;
+
+            // Set threads_max and initial value for threads_use to -1
+            let threads_max = proc::get_threads_max_data(&pids);
+            let mut threads_use = -1_f64;
+
+            let mut interval = interval(Duration::from_millis(2));
+            loop {
+                tokio::select! {
+                    biased;
+                    // When the stop signal is received, generate final metrics
+                    _ = send_metrics.changed() => {
+                        let mut aggr: std::sync::MutexGuard<Aggregator> =
+                            aggr.lock().expect("lock poisoned");
+                        Self::generate_fd_enhanced_metrics(fd_max, fd_use, &mut aggr);
```

Comment:
Right, I had it in the struct because it performs similar logic of inserting the metrics as some of the other methods that use self, should I move all of the `generate_*_metrics` functions out of Lambda?